src/Data/Set.purs

```diff
@@ -45,31 +46,49 @@ isEmpty :: forall a. Set a -> Boolean
 isEmpty (Set m) = M.isEmpty m
 
 singleton :: forall a. a -> Set a
-singleton a = Set (M.singleton a P.unit)
-  
+singleton a = Set (M.singleton a unit)
+
 checkValid :: forall a. Set a -> Boolean
 checkValid (Set m) = M.checkValid m
 
-member :: forall a. (P.Ord a) => a -> Set a -> Boolean
+member :: forall a. (Ord a) => a -> Set a -> Boolean
 member a (Set m) = a `M.member` m
 
-insert :: forall a. (P.Ord a) => a -> Set a -> Set a
-insert a (Set m) = Set (M.insert a P.unit m)
-  
-delete :: forall a. (P.Ord a) => a -> Set a -> Set a
+insert :: forall a. (Ord a) => a -> Set a -> Set a
+insert a (Set m) = Set (M.insert a unit m)
+
+delete :: forall a. (Ord a) => a -> Set a -> Set a
 delete a (Set m) = Set (a `M.delete` m)
-  
+
 toList :: forall a. Set a -> [a]
 toList (Set m) = map fst (M.toList m)
 
-fromList :: forall a. (P.Ord a) => [a] -> Set a
+fromList :: forall a. (Ord a) => [a] -> Set a
 fromList = foldl (\m a -> insert a m) empty
 
-union :: forall a. (P.Ord a) => Set a -> Set a -> Set a
+union :: forall a. (Ord a) => Set a -> Set a -> Set a
 union (Set m1) (Set m2) = Set (m1 `M.union` m2)
 
-unions :: forall a. (P.Ord a) => [Set a] -> Set a
+unions :: forall a. (Ord a) => [Set a] -> Set a
 unions = foldl union empty
 
-difference :: forall a. (P.Ord a) => Set a -> Set a -> Set a
-difference s1 s2 = foldl (P.flip delete) s1 (toList s2)
+difference :: forall a. (Ord a) => Set a -> Set a -> Set a
+difference s1 s2 = foldl (flip delete) s1 (toList s2)
+
+-- | True if and only if every element in the first set
+-- | is an element of the second set
+subset :: forall a. (Ord a) => Set a -> Set a -> Boolean
+subset s1 s2 = isEmpty $ s1 `difference` s2
+
+-- | True if and only if the first set is a subset of the second set
+-- | and the sets are not equal
+properSubset :: forall a. (Ord a) => Set a -> Set a -> Boolean
+properSubset s1 s2 = subset s1 s2 && (s1 /= s2)
+
+-- | The set of elements which are in both the first and second set
+intersection :: forall a. (Ord a) => Set a -> Set a -> Set a
```

Comment:
There has to be a more efficient way to implement this one. Even using `toList` and `fromList` seems like it would probably be faster. Or can we write a `filter` function?